From b6881245a9cbfed0828f2b9b6af13b8b9fe3bb9a Mon Sep 17 00:00:00 2001
From: Vasilis Liaskovitis <vliaskovitis@suse.com>
Date: Sun, 18 Oct 2020 00:46:20 +0200
Subject: [PATCH 07/10] [PATCH 07/11] modesetting: add an option to
 force-enable per-crtc  shadow buffers

This allows us to validate that this codepath is still working
properly, without requiring adding tons of displays and separating
them enough to hit the maximum stride of the display.

Signed-off-by: Martin Peres <martin.peres@linux.intel.com>
---
 hw/xfree86/drivers/modesetting/driver.c          |  8 ++++++++
 hw/xfree86/drivers/modesetting/driver.h          |  1 +
 hw/xfree86/drivers/modesetting/drmmode_display.c |  3 +++
 hw/xfree86/drivers/modesetting/drmmode_display.h |  3 +++
 hw/xfree86/drivers/modesetting/modesetting.man   | 13 +++++++++++++
 5 files changed, 28 insertions(+)

diff --git a/hw/xfree86/drivers/modesetting/driver.c b/hw/xfree86/drivers/modesetting/driver.c
index dee6e5009..b837ac012 100644
--- a/hw/xfree86/drivers/modesetting/driver.c
+++ b/hw/xfree86/drivers/modesetting/driver.c
@@ -133,6 +133,7 @@ static const OptionInfoRec Options[] = {
     {OPTION_ZAPHOD_HEADS, "ZaphodHeads", OPTV_STRING, {0}, FALSE},
     {OPTION_DOUBLE_SHADOW, "DoubleShadow", OPTV_BOOLEAN, {0}, FALSE},
     {OPTION_ATOMIC, "Atomic", OPTV_BOOLEAN, {0}, FALSE},
+    {OPTION_SHADOW_PRIMARY, "ShadowPrimary", OPTV_BOOLEAN, {0}, FALSE},
     {-1, NULL, OPTV_NONE, {0}, FALSE}
 };
 
@@ -1041,6 +1042,13 @@ PreInit(ScrnInfoPtr pScrn, int flags)
         ms->drmmode.sw_cursor = TRUE;
     }
 
+    ms->drmmode.shadow_primary =
+            xf86ReturnOptValBool(ms->drmmode.Options, OPTION_SHADOW_PRIMARY,
+                                 FALSE);
+    if (ms->drmmode.shadow_primary) {
+        xf86DrvMsg(pScrn->scrnIndex, X_INFO, "ShadowPrimary enabled\n");
+    }
+
     ms->cursor_width = 64;
     ms->cursor_height = 64;
     ret = drmGetCap(ms->fd, DRM_CAP_CURSOR_WIDTH, &value);
diff --git a/hw/xfree86/drivers/modesetting/driver.h b/hw/xfree86/drivers/modesetting/driver.h
index f2e7889db..1412fe7e7 100644
--- a/hw/xfree86/drivers/modesetting/driver.h
+++ b/hw/xfree86/drivers/modesetting/driver.h
@@ -52,6 +52,7 @@ typedef enum {
     OPTION_ZAPHOD_HEADS,
     OPTION_DOUBLE_SHADOW,
     OPTION_ATOMIC,
+    OPTION_SHADOW_PRIMARY,
 } modesettingOpts;
 
 typedef struct
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.c b/hw/xfree86/drivers/modesetting/drmmode_display.c
index f368e04e0..3b58b470e 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.c
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.c
@@ -1636,6 +1636,9 @@ drmmode_need_shadow_scanout(xf86CrtcPtr crtc) {
     if (crtc->rotation != RR_Rotate_0)
         return FALSE;
 
+    if (drmmode->shadow_primary)
+        return TRUE;
+
     /* check if the current screen size exceeds the driver's limits */
     mode_res = drmModeGetResources(drmmode->fd);
     if (!mode_res)
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.h b/hw/xfree86/drivers/modesetting/drmmode_display.h
index 73ceced16..f47b05208 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.h
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.h
@@ -104,6 +104,9 @@ typedef struct {
     Bool glamor;
     Bool shadow_enable;
     Bool shadow_enable2;
+    /** force creating a per-crtc framebuffer **/
+    Bool shadow_primary;
+
     /** Is Option "PageFlip" enabled? */
     Bool pageflip;
     Bool force_24_32;
diff --git a/hw/xfree86/drivers/modesetting/modesetting.man b/hw/xfree86/drivers/modesetting/modesetting.man
index d530d7c4c..abf450c7c 100644
--- a/hw/xfree86/drivers/modesetting/modesetting.man
+++ b/hw/xfree86/drivers/modesetting/modesetting.man
@@ -81,6 +81,19 @@ For example:
 Option \*qZaphodHeads\*q \*qLVDS,VGA-0\*q
 will assign xrandr outputs LVDS and VGA-0 to this instance of the driver.
 .TP
+.BI "Option \*qShadowPrimary\*q \*q" boolean \*q
+This option enables a so-called "shadow primary" buffer for fast CPU access to
+pixel data, and separate scanout buffers for each display controller (CRTC).
+This may improve performance for some 2D workloads, potentially at the expense
+of other (e.g. 3D, video) workloads. This option may be enabled automatically
+when the root window size exceeds the maximum stride supported by the display
+controler of your GPU. It is recommended to use the XRender backend of your
+compositor when enabling this feature.
+.br
+In an effort to reduce the performance cost, the latency may be negatively
+affected. The default is
+.B off.
+.TP
 .SH "SEE ALSO"
 @xservername@(@appmansuffix@), @xconfigfile@(@filemansuffix@), Xserver(@appmansuffix@),
 X(@miscmansuffix@)
-- 
2.25.1

